diff --git a/xapian-core/queryparser/queryparser.lemony b/xapian-core/queryparser/queryparser.lemony
index 8dedb80..da1a69f 100644
--- a/xapian-core/queryparser/queryparser.lemony
+++ b/xapian-core/queryparser/queryparser.lemony
@@ -1838,6 +1838,7 @@ prob_expr(E) ::= prob(P). {
     if (P->hate && !P->hate->empty()) {
 	if (!E) {
 	    // Can't just hate!
+	    state->error = "Syntax Error: Can't just HATE a term.\n Did you meant to use the HATE query? \nIf not, try removing '-'or giving space after '-'";
 	    yy_parse_failed(yypParser);
 	    return;
 	}
